support-frontend/app/controllers/RedemptionController.scala

```diff
@@ -0,0 +1,147 @@
+import controllers.RedemptionController._
+import io.circe.syntax._
+import play.api.libs.circe.Circe
+import play.api.mvc.{AbstractController, Action, AnyContent, ControllerComponents, Result}
+import services.{IdentityService, TestUserService}
+import views.EmptyDiv
+import views.html.helper.CSRF
+import views.html.subscriptionRedemptionForm
+import cats.implicits._
+import lib.RedirectWithEncodedQueryString
+import play.twirl.api.Html
+import views.ViewHelpers.outputJson
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class RedemptionController(
+  val actionRefiners: CustomActionBuilders,
+  val assets: AssetsResolver,
+  settingsProvider: AllSettingsProvider,
+  identityService: IdentityService,
+  testUsers: TestUserService,
+  components: ControllerComponents,
+  fontLoaderBundle: Either[RefPath, StyleContent]
+)(
+  implicit val ec: ExecutionContext
+) extends AbstractController(components) with Circe {
+
+  import actionRefiners._
+
+  implicit val a: AssetsResolver = assets
+
+  implicit val settings: AllSettings = settingsProvider.getAllSettings()
+  val title = "Support the Guardian | Redeem your code"
+  val id = EmptyDiv("subscriptions-redemption-page")
+  val js = "subscriptionsRedemptionPage.js"
+  val css = "digitalSubscriptionCheckoutPage.css" //TODO: Don't need this?
+
+  def displayForm(redemptionCode: String): Action[AnyContent] = NoCacheAction().async {
+    implicit request =>
+      getCorporateCustomer(redemptionCode).value.map(
+        customerOrError =>
+          Ok(subscriptionRedemptionForm(
+            title = title,
+            mainElement = id,
+            js = js,
+            css = css,
+            fontLoaderBundle = fontLoaderBundle,
+            csrf = None,
+            uatMode = false,
+            stage = "checkout",
+            redemptionCode = redemptionCode,
+            customerOrError = customerOrError,
+            user = None,
+            submitted = false
+          ))
+      )
+
+  }
+
+  def displayThankYou(stage: String): Action[AnyContent] = authenticatedAction(subscriptionsClientId) {
+    implicit request =>
+          Ok(views.html.main(
+            title = title,
+            mainElement = id,
+            mainJsBundle = Left(RefPath(js)),
+            mainStyleBundle = Left(RefPath(css)),
+            fontLoaderBundle = fontLoaderBundle,
+          ){
+            Html(s"""<script type="text/javascript">window.guardian.stage = "${stage}";</script>""")
+          }
+      )
+  }
+
+  def displayProcessing(redemptionCode: String): Action[AnyContent] =
+    authenticatedAction(subscriptionsClientId).async {
+      implicit request: AuthRequest[Any] =>
+        val processingPage = for {
+          user <- identityService.getUser(request.user.minimalUser)
+          corporateCustomer <- getCorporateCustomer(redemptionCode)
+        } yield showProcessing(redemptionCode, corporateCustomer, user)
+
+        processingPage.value.map(
+          _.fold(
+            BadRequest(_),
```

Comment:
this will potentially send the identity url or any json parse error of the identity response back to the client which is not ideal for security.
The identityService should probably use failed future for any unexpected problem with the call, and Left should be reserved for signalling something specific e.g. user not found.

Reply:
Have changed the way this works